### PR TITLE
SmartSocket implementation

### PR DESCRIPTION
**Pull request by vsh51**

### CODE REVIEW

**Review by chrkspln — APPROVED** [review]
code is well-structured, nicely implements boost's lib features, looks clean and up to code convention. approved

**Review by TheCheezyPie — APPROVED** [review]
I will agree with my colleague on this, I have nothing to add regarding the code quality (it's nice).

Although I have found some really small problems that have _**no real impact**_:

1. By the class format rule, the data members should be the last declarations. The SmartSocket::StartTimer function is the only breaker of this rule here.
2. The colon in the constructor initializer list should be on the new line.
3. The contents of namespaces should not be indented. (although many IDEs automatically discard this rule so this may not be the code author's fault)

File: libs/SmartSocket/include/SmartSocket.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+#include <regex>
+#include <future>
+
+using namespace boost;
+
+using std::string;
+using asio::ip::tcp;
+using std::function;
+
+namespace ISXSmartSocket
```
Comment (TheCheezyPie) [inline libs/SmartSocket/include/SmartSocket.h:20]: Namespace contents should not be indented

File: libs/SmartSocket/include/SmartSocket.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        bool m_ssl_enabled;
+
+        int m_timeout = 5;
+
+        tcp::resolver m_resolver;
+        asio::io_context& m_io_context;
+        asio::ssl::context& m_ssl_context;
+        asio::ssl::stream<tcp::socket> m_socket;
+        
+        std::unique_ptr<asio::steady_timer> StartTimer(
```
Comment (TheCheezyPie) [inline libs/SmartSocket/include/SmartSocket.h]: data members declarations should be last

File: libs/SmartSocket/src/SmartSocket.cpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+#include <iostream>
+#include <boost/asio/steady_timer.hpp>
+#include <memory>
+
+namespace ISXSmartSocket
+{
+    SmartSocket::SmartSocket(
+        asio::io_context& io_context
+        , asio::ssl::context& ssl_context):
```
Comment (TheCheezyPie) [inline libs/SmartSocket/src/SmartSocket.cpp]: colon should be on the new line

File: libs/SmartSocket/src/SmartSocket.cpp [hunk 4]
```diff
@@ -0,0 +1,149 @@
+#include "SmartSocket.h"
+#include "Handlers.h"
+
+#include <iostream>
+#include <boost/asio/steady_timer.hpp>
+#include <memory>
+
+namespace ISXSmartSocket
```
Comment (TheCheezyPie) [inline libs/SmartSocket/src/SmartSocket.cpp:8]: Namespace contents should not be indented

**Comment by vsh51** [comment]
There were two dependency libraries added to show basic functionality. "Handlers" and "SMTPResponse" libs are under review on other pull requests, so there is no need to examine them here. They will be updated later

**Pull request closed without merging**